src/pmdas/amdgpu/amdgpu.c

```diff
@@ -0,0 +1,668 @@
+      amdgpu_info_t *info = &amdgpuinfo->info[i];
+      void *param = NULL;
+      int ret = DRM_SUCCESS;
+
+      switch (to_refresh) {
+	case AMDGPU_NAME_REFRESHER:
+	  param = info->name;
+	  break;
+	case AMDGPU_GPU_INFO_REFRESHER:
+	  param = &info->gpu_info;
+	  break;
+	case AMDGPU_GPU_CLOCK_REFRESHER:
+	  param = &info->gpu_clock;
+	  break;
+	case AMDGPU_GPU_TEMPERATURE_REFRESHER:
+	  param = &info->temperature;
+	  break;
+	case AMDGPU_GPU_LOAD_REFRESHER:
+	  param = &info->load;
+	  break;
+	case AMDGPU_GPU_AVG_PWR_REFRESHER:
+	  param = &info->avg_pwr;
+	  break;
+	case AMDGPU_MEMORY_INFO_REFRESHER:
+	  param = &memory;
+	  break;
+	case AMDGPU_MEMORY_CLOCK_REFRESHER:
+	  param = &info->mem_clock;
+	  break;
+      }
+
+      ret = amd_refresher[to_refresh].refresher(info->amd_device, param);
+
+      if (ret != DRM_SUCCESS) {
+	  /* Mark all metrics that depend on the same refresher as failed */
+	  for (int j = 0; j < CLUSTER_COUNT; j++)
+	    for (int k = 0; k < MAX_ITEM_COUNT; k++) {
+		if (&amd_refresher[to_refresh] == refresher_list[j][k])
+		  info->failed[j][k] = 1;
+	    }
+
+	  continue;
+      }
+
+      if (param == &memory) {
+	  info->memory = memory; /* struct copy */
+	  info->memaccum += memory.used;
+	  pmNotifyErr(LOG_ERR, "Refreshed memory (%lx)", memory.used);
+      }
+  }
+
+  return 0;
+}
+
+static int amdgpu_instance(pmInDom indom, int inst, char *name,
+                           pmInResult **result, pmdaExt *pmda) {
+  return pmdaInstance(indom, inst, name, result, pmda);
+}
+
+/*
+ * Wrapper for pmdaFetch which refresh the set of values once per fetch
+ * PDU.  The fetchCallback is then called once per-metric/instance pair
+ * to perform the actual filling of the pmResult (via each pmAtomValue).
+ */
+static int amdgpu_fetch(int numpmid, pmID pmidlist[], pmResult **resp,
+                        pmdaExt *pmda)
+{
+  uint32_t i = 0;
+
+  pmNotifyErr(LOG_ERR, "amdgpu_fetch");
+
+  for (i = 0; i <= numpmid; i++) {
+      unsigned int cluster = pmID_cluster(pmidlist[i]);
+      unsigned int item = pmID_item(pmidlist[i]);
+
+      pmNotifyErr(LOG_ERR, "amdgpu_fetch (%d/%d/%d)", cluster, item, numpmid);
+
+      if (cluster >= CLUSTER_COUNT)
+	continue;
+      if (refresher_list[cluster][item] == NULL)
+	continue;
+
+      refresher_list[cluster][item]->needs_refresh = 1;
+  }
+
+  for (i = 0; i < AMDGPU_REFRESHER_COUNT; i++) {
+      if (!amd_refresher[i].needs_refresh)
+	continue;
+      amd_refresher[i].needs_refresh = 0;
+      refresh(&pcp_amdgpuinfo, i);
+      pmNotifyErr(LOG_ERR, "amdgpu_fetch refresh %d", i);
+  }
+
+  return pmdaFetch(numpmid, pmidlist, resp, pmda);
+}
+
+static int amdgpu_fetchCallBack(pmdaMetric *mdesc, unsigned int inst,
+                                pmAtomValue *atom) {
+  unsigned int cluster = pmID_cluster(mdesc->m_desc.pmid);
+  unsigned int item = pmID_item(mdesc->m_desc.pmid);
+
+  pmNotifyErr(LOG_ERR, "amdgpu_fetchCallBack (%d/%d/%d)", cluster, item, inst);
```

Comment:
Too verbose - perhaps under "if (pmDebugOptions.xxx)" conditional?  Run pmdbg(1) --list for options.